File: samples/opticalflow_cuda/java/OpticalFlowCuda.java
import org.bytedeco.javacpp.indexer.FloatIndexer;
import org.bytedeco.opencv.global.opencv_imgproc;
import org.bytedeco.opencv.global.opencv_videoio;
import org.bytedeco.opencv.opencv_core.GpuMat;
import org.bytedeco.opencv.opencv_core.Mat;
import org.bytedeco.opencv.opencv_core.Point;
import org.bytedeco.opencv.opencv_core.Scalar;
import org.bytedeco.opencv.opencv_cudaoptflow.FarnebackOpticalFlow;
import org.bytedeco.opencv.opencv_videoio.VideoCapture;

public class OpticalFlowCuda {
    public static void main(String[] args) {
        int width = 640;
        int height = 480;
        VideoCapture cap = new VideoCapture();
        cap.set(opencv_videoio.CAP_PROP_FRAME_WIDTH, width);
        cap.set(opencv_videoio.CAP_PROP_FRAME_HEIGHT, height);
        cap.open(0);
        GpuMat prevgray_cuda = null;
        FarnebackOpticalFlow optflow_cuda = FarnebackOpticalFlow.create();
        while (true) {
            Mat frame = new Mat();
            boolean ret = cap.read(frame);
            if (!ret) break;

            // gray scale
            Mat gray = new Mat();
            opencv_imgproc.cvtColor(frame, gray, opencv_imgproc.COLOR_BGR2GRAY);

            if (prevgray_cuda == null || prevgray_cuda.empty()) {
                prevgray_cuda = new GpuMat(gray);
                releaseMat(frame, gray);
                continue;
            }

            // calculate dense optical flow
            GpuMat gray_cuda = new GpuMat(gray);
            GpuMat flow_cuda = new GpuMat();
            Mat flow = new Mat();
            optflow_cuda.calc(prevgray_cuda, gray_cuda, flow_cuda);
            flow_cuda.download(flow);

            // draw direction of flow
            FloatIndexer flowatxy = flow.createIndexer();
            for (int j = 0; j < frame.rows(); j += 20) {
                for (int i = 0; i < frame.cols(); i += 20) {
                    float fi = flowatxy.get(j, i, 0);
                    float fj = flowatxy.get(j, i, 1);
                    opencv_imgproc.line(frame, new Point(i, j),
                            new Point((int) (i + fi), (int) (j + fj)),
                            Scalar.all(255), 2, opencv_imgproc.LINE_AA, 0);
                    opencv_imgproc.line(frame, new Point(i, j),
                            new Point((int) (i + fi), (int) (j + fj)),
                            new Scalar(192, 0, 0, 0), 1, opencv_imgproc.LINE_AA, 0);
                    opencv_imgproc.circle(frame, new Point(i, j), 1,
                            Scalar.all(255), 2, opencv_imgproc.LINE_AA, 0);
                    opencv_imgproc.circle(frame, new Point(i, j), 1,
                            new Scalar(0, 133, 255, 0), 1, opencv_imgproc.LINE_AA, 0);
                }
            }

            OpenCVFX.imshow("Dense Optical Flow", frame);
            if (OpenCVFX.waitKey() == 27) {
                break;
            }

            prevgray_cuda = gray_cuda;

            // release mat
            releaseMat(frame, gray, flow);
        }
        cap.release();
        OpenCVFX.destroyAllWindows();
    }

    private static void releaseMat(Mat... mats) {
        for (Mat mat : mats) {
            if (mat != null) {
                mat.release();
                mat.close();
            }
        }
    }

    private static void releaseGpuMat(GpuMat... mats) {
        for (GpuMat mat : mats) {
            if (mat != null) {
                mat.release();
                mat.close();
            }
        }
    }
}

File: samples/opticalflow_cuda/python/opticalflow_cuda.py
import cv2
import numpy as np

width = 640
height = 480
cap = cv2.VideoCapture()
cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
cap.open(0)
prevgray_cuda = None

optflow_cuda = cv2.cuda.FarnebackOpticalFlow_create(numLevels=5, pyrScale=0.5, fastPyramids=False,
    winSize=13, numIters=10, polyN=5, polySigma=1.1, flags=0)

while True:
    ret, frame = cap.read()
    if not ret:
        break

    # gray scale
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    if prevgray_cuda is None:
        prevgray_cuda = cv2.cuda_GpuMat(gray.copy())
        continue

    # calculate dense optical flow
    gray_cuda = cv2.cuda_GpuMat()
    gray_cuda.upload(gray)
    flow_cuda = optflow_cuda.calc(prevgray_cuda, gray_cuda, None)
    flow = flow_cuda.download()

    # draw direction of flow
    for y in range(0, frame.shape[0], 20):
        for x in range(0, frame.shape[1], 20):
            flowatxy = flow[y, x]
            cv2.line(frame, (x, y), (int(
                x+flowatxy[0]), int(y+flowatxy[1])), (255, 255, 255), 2, cv2.LINE_AA)
            cv2.line(frame, (x, y), (int(
                x+flowatxy[0]), int(y+flowatxy[1])), (192, 0, 0), 1, cv2.LINE_AA)
            cv2.circle(frame, (x, y), 1, (255, 255, 255), 2, cv2.LINE_AA)
            cv2.circle(frame, (x, y), 1, (0, 133, 255), 1, cv2.LINE_AA)

    cv2.imshow("Dense Optical Flow", frame)

    key = cv2.waitKey(10)
    if key == 27:
        break

    prevgray_cuda = gray_cuda


cap.release()
cv2.destroyAllWindows()
